compiler/rustc_builtin_macros/src/concat_bytes.rs

```diff
@@ -0,0 +1,210 @@
+    let mut has_errors = false;
+    for e in es {
+        match e.kind {
+            ast::ExprKind::Array(ref exprs) => {
+                for expr in exprs {
+                    match expr.kind {
+                        ast::ExprKind::Array(_) => {
+                            cx.span_err(expr.span, "cannot concatenate doubly nested array");
+                            has_errors = true;
+                        }
+                        ast::ExprKind::Lit(ref lit) => match lit.kind {
+                            ast::LitKind::Str(_, _) => {
+                                cx.span_err(
+                                    expr.span,
+                                    "cannot concatenate an array of string literals",
+                                );
+                                has_errors = true;
+                            }
+                            ast::LitKind::Float(_, _) => {
+                                cx.span_err(
+                                    expr.span,
+                                    "cannot concatenate an array of float literals",
+                                );
+                                has_errors = true;
+                            }
+                            ast::LitKind::Char(_) => {
+                                if let Ok(snippet) = cx.sess.source_map().span_to_snippet(expr.span)
+                                {
+                                    cx.struct_span_err(
+                                        expr.span,
+                                        "cannot concatenate an array of character literals",
+                                    )
+                                    .span_suggestion(
+                                        expr.span,
+                                        "try using a byte character",
+                                        format!("b{}", snippet),
+                                        Applicability::MachineApplicable,
+                                    )
+                                    .emit();
+                                } else {
+                                    cx.span_err(
+                                        expr.span,
+                                        "cannot concatenate an array of character literals",
+                                    )
+                                }
+                                has_errors = true;
+                            }
+                            ast::LitKind::Int(
+                                val,
+                                ast::LitIntType::Unsuffixed
+                                | ast::LitIntType::Unsigned(ast::UintTy::U8),
+                            ) if val <= 255 => {
+                                accumulator.push(val as u8);
+                            }
+                            ast::LitKind::Int(_, _) => {
+                                cx.span_err(expr.span, "numeric literal is out of bounds");
```

Comment:
I think this will give an out of bounds error on e.g. `5u16`, which feels a bit unfortunate: can we provide a dedicated error for specifically the non-U8 case, regardless of the value?